fem/bilininteg_ext.cpp

```diff
@@ -378,294 +371,376 @@ void PADiffusionApply2D(const int NE,
 {
    const int D1D = T_D1D ? T_D1D : d1d;
    const int Q1D = T_Q1D ? T_Q1D : q1d;
-   MFEM_VERIFY(D1D <= MAX_D1D, "");
-   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
-
-   const DeviceMatrix B(b, Q1D, D1D);
-   const DeviceMatrix G(g, Q1D, D1D);
-   const DeviceMatrix Bt(bt, D1D, Q1D);
-   const DeviceMatrix Gt(gt, D1D, Q1D);
+   const int NBZ = T_NBZ ? T_NBZ : 1;
+   const int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+   const int MD1 = T_D1D ? T_D1D : MAX_D1D;
+   MFEM_VERIFY(MQ1 == MD1, "");
+   MFEM_VERIFY(D1D <= MD1, "");
+   MFEM_VERIFY(Q1D <= MQ1, "");
+   const DeviceMatrix b(_b, Q1D, D1D);
+   const DeviceMatrix g(_g, Q1D, D1D);
+   const DeviceMatrix bt(_bt, D1D, Q1D);
+   const DeviceMatrix gt(_gt, D1D, Q1D);
    const DeviceTensor<3> op(_op, 3, Q1D*Q1D, NE);
    const DeviceTensor<3> x(_x, D1D, D1D, NE);
    DeviceTensor<3> y(_y, D1D, D1D, NE);
-
-   MFEM_FORALL(e, NE,
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, NBZ,
    {
       const int D1D = T_D1D ? T_D1D : d1d; // nvcc workaround
       const int Q1D = T_Q1D ? T_Q1D : q1d;
-
-      double grad[MAX_Q1D][MAX_Q1D][2];
-      for (int qy = 0; qy < Q1D; ++qy)
+      const int NBZ = T_NBZ ? T_NBZ : 1;
+      const int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+      const int MD1 = T_D1D ? T_D1D : MAX_D1D;
+      MFEM_SHARED double B[MQ1][MD1];
+      MFEM_SHARED double G[MQ1][MD1];
+      MFEM_SHARED double Bt[MD1][MQ1];
+      MFEM_SHARED double Gt[MD1][MQ1];
+      MFEM_SHARED double Xz[NBZ][MD1][MD1];
+      MFEM_SHARED double GD[2][NBZ][MD1][MQ1];
+      MFEM_SHARED double GQ[2][NBZ][MD1][MQ1];
```

Comment:
For high enough orders it won't make sense to store all of these in shared memory since it will destroy your occupancy.  Do we have a non-shared memory variant that we can call in that case?  It also might make sense to have a variant that has just B and G in shared memory for the in between case.

Reply:
I'll bring back the original ones with no shared memory to use them for the cases you're describing.